### DISCUSSION TITLE
the use of threading.lock() in the performance example

### DISCUSSION THREAD

**Discussion by alexandrecuer** [pymodbus-dev/pymodbus#1269] [General]
I can see that the performance example has been reworked 
https://github.com/pymodbus-dev/pymodbus/blob/dev/examples/v2.5.3/performance.py

you do not use anymore the threading.lock() in the test, even if the variable is still in the code 
https://github.com/pymodbus-dev/pymodbus/blob/ab83cc8487b263586842ecdb5be1b8929bae7f06/examples/v2.5.3/performance.py#L34

I guess it is because the test covers modbus tcp ?

If testing in modbus rtu,the lock would be necessary ?

Sorry if I am misunderstanding things...

best

Alex

**Comment by janiversen** [comment]
I am not aware that is has been reworked, that is basically the reason why it is in the v2.5.3 directory. When reworked it will move up to the other updated examples.

In version 3.x we do not use locks in tests.

**Comment by alexandrecuer** [comment]
Thanks
Can i just ask why it was decided not to use locks ?
In practise in rtu mode, we need to assure at if a thread wants to send a second request, it will wait that the previous one, sent by another thread, has been answered. So yu need a lock or a queue, unless you work on modbus tcp which accepts multiple connections.

**Comment by janiversen** [comment]
The code was rewritten (not the example) and I am sure when the example gets rewritten it will be without a lock(). That is not the same as to say applications never need to use locks, just that the performance example do not need it.

**Comment by alexandrecuer** [comment]
by the way, there is a lock in the transaction manager, where the request is processed, if I am correct....
https://github.com/pymodbus-dev/pymodbus/blob/94fa6b1918df2f7d104ccfe1893ffe229cd142e0/pymodbus/transaction.py#L63
https://github.com/pymodbus-dev/pymodbus/blob/94fa6b1918df2f7d104ccfe1893ffe229cd142e0/pymodbus/transaction.py#L124

used in the base client
https://github.com/pymodbus-dev/pymodbus/blob/94fa6b1918df2f7d104ccfe1893ffe229cd142e0/pymodbus/client/base.py#L134

I need more time to understand things, asyncio is not so easy to decrypt for me

**Reply by janiversen** [reply]
I was not talking about what we do internally in the code, but what was needed to make performance.py work.
